### PR TITLE
add basic support & tests for lmpop

### PR DESCRIPTION
**Pull request by funcmike**
Solves #273

### CODE REVIEW

**Comment by funcmike** [comment]
@microsoft-github-policy-service agree

**Review by TalZaccai — CHANGES_REQUESTED** [review]

File: libs/server/Resp/Objects/ListCommands.cs [hunk 1]
```diff
@@ -183,6 +183,108 @@ command switch
...
 
+
+        /// <summary>
+        /// LMPOP numkeys key [key ...] LEFT | RIGHT [COUNT count]
+        /// </summary>
+        /// <param name="count"></param>
+        /// <param name="ptr"></param>
+        /// <param name="storageApi"></param>
+        /// <returns></returns>
+        private unsafe bool ListManyPop<TGarnetApi>(int count, byte* ptr, ref TGarnetApi storageApi)
```
[the remark this answers is not in the document]
Comment (funcmike) [reply libs/server/Resp/Objects/ListCommands.cs]: renamed

File: libs/server/Resp/Objects/ListCommands.cs [hunk 2]
```diff
@@ -183,6 +183,108 @@ command switch
...
+                            where TGarnetApi : IGarnetApi
+        {
+            if (count < 3)
+            {
+                return AbortWithWrongNumberOfArguments("LMPOP", count);
+            }
+
+            int numKeys = 1;
```
[the remark this answers is not in the document]
Comment (funcmike) [reply libs/server/Resp/Objects/ListCommands.cs]: changed

File: libs/server/Resp/Objects/ListCommands.cs [hunk 3]
```diff
@@ -183,6 +183,108 @@ command switch
...
+
+            // Read count of keys
+            if (!RespReadUtils.ReadIntWithLengthHeader(out numKeys, ref ptr, recvBufferPtr + bytesRead))
+                return false;
+
+            // Get the keys for Lists
+            var keys = new ArgSlice[numKeys];
+
```
[the remark this answers is not in the document]
Comment (funcmike) [reply libs/server/Resp/Objects/ListCommands.cs:213]: check added

File: libs/server/Resp/Objects/ListCommands.cs [hunk 4]
```diff
@@ -183,6 +183,108 @@ command switch
...
+            for (var i = 0; i < keys.Length; i++)
+            {
+                keys[i] = default;
+                if (!RespReadUtils.ReadPtrWithLengthHeader(ref keys[i].ptr, ref keys[i].length, ref ptr, recvBufferPtr + bytesRead))
+                    return false;
+            }
+
+            if (NetworkMultiKeySlotVerify(readOnly: false, firstKey: 0, lastKey: -2))
```
[the remark this answers is not in the document]
Comment (funcmike) [reply libs/server/Resp/Objects/ListCommands.cs]: corrected

File: libs/server/Resp/Objects/ListCommands.cs [hunk 5]
```diff
@@ -183,6 +183,108 @@ command switch
...
+                return true;
+
+            ArgSlice dir = default;
+
+            if (!RespReadUtils.ReadPtrWithLengthHeader(ref dir.ptr, ref dir.length, ref ptr, recvBufferPtr + bytesRead))
+                return false;
+
+            var popDirection = GetOperationDirection(dir);
```
[the remark this answers is not in the document]
Comment (funcmike) [reply libs/server/Resp/Objects/ListCommands.cs:229]: check added

File: libs/server/Storage/Session/ObjectStore/ListOps.cs [hunk 6]
```diff
@@ -145,6 +145,49 @@ public unsafe GarnetStatus ListPop<TObjectContext>(ArgSlice key, int count, List
...
+        /// <typeparam name="TObjectContext"></typeparam>
+        /// <param name="keys"></param>
+        /// <param name="direction"></param>
+        /// <param name="count"></param>
+        /// <param name="context"></param>
+        /// <param name="objectContext"></param>
+        /// <param name="key"></param>
+        /// <param name="elements"></param>
+        /// <returns>The count elements popped from the list</returns>
+        public unsafe GarnetStatus ListManyPop<TContext, TObjectContext>(ArgSlice[] keys, OperationDirection direction, int count, ref TContext context, ref TObjectContext objectContext, out ArgSlice key, out ArgSlice[] elements)
```
[the remark this answers is not in the document]
Comment (funcmike) [reply libs/server/Storage/Session/ObjectStore/ListOps.cs]: renamed

File: libs/server/Storage/Session/ObjectStore/ListOps.cs [hunk 7]
```diff
@@ -145,6 +145,49 @@ public unsafe GarnetStatus ListPop<TObjectContext>(ArgSlice key, int count, List
...
+            where TContext : ITsavoriteContext<SpanByte, SpanByte, SpanByte, SpanByteAndMemory, long, MainStoreFunctions>
+            where TObjectContext : ITsavoriteContext<byte[], IGarnetObject, SpanByte, GarnetObjectStoreOutput, long, ObjectStoreFunctions>
+        {
+            foreach (var k in keys )
+            {
+                if (EXISTS(k, StoreType.All, ref context, ref objectContext) != GarnetStatus.OK) continue;
```
[the remark this answers is not in the document]
Comment (funcmike) [reply libs/server/Storage/Session/ObjectStore/ListOps.cs]: removed

File: libs/server/Storage/Session/ObjectStore/ListOps.cs [hunk 8]
```diff
@@ -145,6 +145,49 @@ public unsafe GarnetStatus ListPop<TObjectContext>(ArgSlice key, int count, List
...
+
+                key = k;
```
[the remark this answers is not in the document]
Comment (funcmike) [reply libs/server/Storage/Session/ObjectStore/ListOps.cs]: corrected

File: libs/server/API/GarnetApiObjectCommands.cs [hunk 9]
```diff
@@ -190,6 +190,10 @@ public unsafe GarnetStatus ListLeftPop(ArgSlice key, out ArgSlice element)
         public GarnetStatus ListLeftPop(ArgSlice key, int count, out ArgSlice[] poppedElements)
```
[the remark this answers is not in the document]
Comment (funcmike) [reply libs/server/API/GarnetApiObjectCommands.cs:190]: 1. done
2. please check if conversion of first arg to int is ok? (maybe there is better way?)

File: test/Garnet.test/RespListTests.cs [hunk 10]
```diff
@@ -1302,6 +1302,35 @@ public void CheckEmptyListKeyRemoved()
             Assert.IsFalse(keyExists);
         }
 
+        [Test]
+        public void CanDoBasicLMPOP()
+        {
```
[the remark this answers is not in the document]
Comment (funcmike) [reply test/Garnet.test/RespListTests.cs:1307]: added

**Review by TalZaccai — CHANGES_REQUESTED** [review]
Thanks for the prompt corrections! Added a few more small comments.

**Comment by badrishc** [comment]
Looking great, thanks @funcmike for the contribution!

File: libs/server/Transaction/TxnKeyManager.cs [hunk 11]
```diff
@@ -296,6 +297,26 @@ private int ListKeys(int inputCount, bool isObject, LockType type)
...
 
+        /// <summary>
+        /// Returns a list of keys for LMPOP command
+        /// </summary>
+        private int LMPOPKeys(int inputCount, bool isObject, LockType type)
+        {
+            var numKeysArg = respSession.GetCommandAsArgSlice(out bool success);
+            if (!success) return -2;
+
+            var numKeys = Convert.ToInt32(numKeysArg.ToString());
```
[the remark this answers is not in the document]
Comment (funcmike) [reply libs/server/Transaction/TxnKeyManager.cs]: changed

File: libs/server/Transaction/TxnKeyManager.cs [hunk 12]
```diff
@@ -296,6 +297,26 @@ private int ListKeys(int inputCount, bool isObject, LockType type)
             return inputCount;
         }
 
+        /// <summary>
+        /// Returns a list of keys for LMPOP command
+        /// </summary>
+        private int LMPOPKeys(int inputCount, bool isObject, LockType type)
```
[the remark this answers is not in the document]
Comment (funcmike) [reply libs/server/Transaction/TxnKeyManager.cs]: converted

File: libs/server/Storage/Session/ObjectStore/ListOps.cs [hunk 13]
```diff
@@ -145,6 +145,47 @@ public unsafe GarnetStatus ListPop<TObjectContext>(ArgSlice key, int count, List
             return status;
         }
 
+        /// <summary>
+        /// Removes the count elements from the head(left) or tail(right) of the first non-empty list key from the list of provided key names.
+        /// If the list contains less than count elements, removes and returns the number of elements in the list.
+        /// </summary>
+        /// <typeparam name="TContext"></typeparam>
```
[the remark this answers is not in the document]
Comment (funcmike) [reply libs/server/Storage/Session/ObjectStore/ListOps.cs]: corrected

File: libs/server/Resp/Objects/ListCommands.cs [hunk 14]
```diff
@@ -183,6 +183,119 @@ command switch
...
+            int popCount = 1;
+
+            if (count == numKeys + 4)
+            {
+                ArgSlice countArg = default;
+
+                if (!RespReadUtils.ReadPtrWithLengthHeader(ref countArg.ptr, ref countArg.length, ref ptr, recvBufferPtr + bytesRead))
+                    return false;
+
+                if (!countArg.ReadOnlySpan.SequenceEqual("COUNT"u8))
```
[the remark this answers is not in the document]
Comment (funcmike) [reply libs/server/Resp/Objects/ListCommands.cs]: replaced

**Review by TalZaccai — APPROVED** [review]
Looks good! Thanks!

**Pull request merged**